<a href="https://colab.research.google.com/github/Bonjiang/datasci_2_manipulation/blob/main/507_HW_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

### Load data

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
df

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

### Identify and Handling Missing Values in a Dataset

In [ ]:
df.replace('missing', np.nan, inplace=True)

In [ ]:
missing_values = df.isnull().sum()
missing_values

Patient Age              5249
Gender                   5261
City of Residence        5227
State of Residence       5248
Has Insurance            5246
Visited Last Month       5245
Payment Method           5202
Preferred Doctor         5247
Disease Diagnosed        5271
Medication Prescribed    5262
Type of Appointment      5250
Average Heart Rate       5278
Average BP               5255
Height (in cm)           5244
Weight (in kg)           5264
Payment Due ($)          5229
Last Visit (days ago)    5259
Visit Duration (mins)    5276
Number of Tests          5272
Prescription Cost ($)    5216
dtype: int64

##### Checking for a specific column

In [ ]:
df['Visited Last Month'].isnull().sum()

5245

##### Dropping rows with any missing values

In [ ]:
df.dropna(inplace=True)
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

 ##### *67406 rows were removed*

### Removing Duplicates

##### Removing duplicate rows

In [ ]:
df.drop_duplicates(inplace=True)
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

##### *1766 rows were removed*

##### Removing duplicate columns

In [ ]:
df = df.T.drop_duplicates().T
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

##### *0 columns were removed*

### Cleaning column names


##### Removing leading and trailing white space from a column

In [ ]:
df.columns = df.columns.str.strip()
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

##### Converting column names to uppercase

In [ ]:
df.columns = df.columns.str.upper()
df

PATIENT AGE  GENDER CITY OF RESIDENCE STATE OF RESIDENCE HAS INSURANCE  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       VISITED LAST MONTH PAYMENT METHOD PREFERRED DOCTOR DISEASE DIAGNOSED  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       MEDICATION PRESCRIBED TYPE OF APPOINTMENT AVERAGE HEART RATE  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       AVERAGE BP HEIGHT (IN CM) WEIGHT (IN KG)     PAYMENT DUE ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       LAST VISIT (DAYS AGO) VISIT DURATION (MINS) NUMBER OF TESTS  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

### Creating New Columns Based on Existing Ones

##### Creating a 'Birth Date' column based on patient age and reference date

In [ ]:
reference_date = '2023-09-12'
df['BIRTH DATE'] = pd.to_datetime(reference_date) - pd.to_timedelta(df['PATIENT AGE'])
df


PATIENT AGE  GENDER CITY OF RESIDENCE STATE OF RESIDENCE HAS INSURANCE  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       VISITED LAST MONTH PAYMENT METHOD PREFERRED DOCTOR DISEASE DIAGNOSED  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       MEDICATION PRESCRIBED  ... AVERAGE HEART RATE AVERAGE BP  \
0                      Med_E  ...                 66        111   
4                      Med_C  ...                 93         89   
7                      Med_E  ...                 99        129   
8                      Med_C  ...                 67        108   
11                     Med_E  ...                 65        106   
...                      ...  ...                ...        ...   
104992                 Med_A  ...                 76         93   
104993                 Med_D  ...                 80        129   
104994                 Med_B  ...                 68        109   
104995                 Med_E  ...                 65        112   
104997                 Med_D  ...                 89        118   

       HEIGHT (IN CM) WEIGHT (IN KG)     PAYMENT DUE ($)  \
0                 183             70   65.45721578126224   
4                 186             87   264.2147372473319   
7                 170             47  387.08504353508863   
8                 173             86   172.5483427968999   
11                166             82   299.1426761400895   
...               ...            ...                 ...   
104992            178             66  462.56567087397013   
104993            171             78  131.35632576282202   
104994            172             75   488.0316041186948   
104995            169             61  309.16444606680665   
104997            179             71   320.8304180033907   

       LAST VISIT (DAYS AGO) VISIT DURATION (MINS) NUMBER OF TESTS  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193                    92               1   
...                      ...                   ...             ...   
104992                   189                    55               3   

### Aggregating Data

In [ ]:
grouped_data = df.groupby('STATE OF RESIDENCE').agg({
    'PATIENT AGE': 'max'
})
grouped_data


PATIENT AGE
STATE OF RESIDENCE            
Alabama                     86
Alaska                      85
Arizona                     94
Arkansas                    97
California                  90
Colorado                    90
Connecticut                 90
Delaware                    87
Florida                     88
Georgia                     95
Hawaii                      85
Idaho                       82
Illinois                    93
Indiana                     95
Iowa                        99
Kansas                      90
Kentucky                    98
Louisiana                   96
Maine                       86
Maryland                    86
Massachusetts               99
Michigan                    88
Minnesota                   89
Mississippi                 94
Missouri                    94
Montana                     96
Nebraska                    93
Nevada                      95
New Hampshire               97
New Jersey                  96
New Mexico                  89
New York                    89
North Carolina              97
North Dakota                90
Ohio                        90
Oklahoma                    92
Oregon                      95
Pennsylvania                95
Rhode Island                93
South Carolina              95
South Dakota                96
Tennessee                   89
Texas                       95
Utah                        98
Vermont                     88
Virginia                    92
Washington                  91
West Virginia               92
Wisconsin                   98
Wyoming                     95

In [ ]:
grouped_data = df.groupby('HAS INSURANCE').agg({
    'VISITED LAST MONTH': 'count'
})
grouped_data


VISITED LAST MONTH
HAS INSURANCE                    
No                          17843
Yes                         17985

### Computing Summary Statistics

In [ ]:
summary_statistics = df.describe()
summary_statistics

<ipython-input-372-9ded3ede0724>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  summary_statistics = df.describe()


PATIENT AGE GENDER CITY OF RESIDENCE STATE OF RESIDENCE HAS INSURANCE  \
count        35828  35828             35828              35828         35828   
unique          83      3             20260                 50             2   
top             18  Other      West Michael          Minnesota           Yes   
freq          1573  12050                40                774         17985   
first          NaN    NaN               NaN                NaN           NaN   
last           NaN    NaN               NaN                NaN           NaN   

       VISITED LAST MONTH PAYMENT METHOD PREFERRED DOCTOR DISEASE DIAGNOSED  \
count               35828          35828            35828             35828   
unique                  2              3                5                 5   
top                    No           Cash     Dr. Williams              Cold   
freq                17980          11996             7274              7228   
first                 NaN            NaN              NaN               NaN   
last                  NaN            NaN              NaN               NaN   

       MEDICATION PRESCRIBED  ... AVERAGE HEART RATE AVERAGE BP  \
count                  35828  ...              35828      35828   
unique                     5  ...                 70         61   
top                    Med_B  ...                 80         80   
freq                    7244  ...               1434        972   
first                    NaN  ...                NaN        NaN   
last                     NaN  ...                NaN        NaN   

       HEIGHT (IN CM) WEIGHT (IN KG)    PAYMENT DUE ($) LAST VISIT (DAYS AGO)  \
count           35828          35828              35828                 35828   
unique             71             87              35828                    76   
top               174             40  65.45721578126224                   184   
freq             1441            975                  1                  1515   
first             NaN            NaN                NaN                   NaN   
last              NaN            NaN                NaN                   NaN   

       VISIT DURATION (MINS) NUMBER OF TESTS PRESCRIPTION COST ($)  \
count                  35828           35828                 35828   
unique                   111              11                 35828   
top                       21               4     15.71751735532889   
freq                     377            9034                     1   
first                    NaN             NaN                   NaN   
last                     NaN             NaN                   NaN   

                           BIRTH DATE  
count                           35828  
unique                             83  
top     2023-09-11 23:59:59.999999982  
freq                             1573  
first   2023-09-11 23:59:59.999999900  
last    2023-09-11 23:59:59.999999982  

[6 rows x 21 columns]

### Using Pivot Tables

In [ ]:
pivot_df = df.pivot_table(index='STATE OF RESIDENCE', columns='GENDER', values='NUMBER OF TESTS', aggfunc='count')
pivot_df


GENDER              Female  Male  Other
STATE OF RESIDENCE                     
Alabama                239   228    249
Alaska                 255   198    243
Arizona                236   252    232
Arkansas               278   201    266
California             254   247    241
Colorado               220   261    246
Connecticut            237   262    257
Delaware               238   232    240
Florida                253   239    213
Georgia                221   253    245
Hawaii                 218   207    242
Idaho                  247   235    253
Illinois               234   232    234
Indiana                215   224    232
Iowa                   244   269    260
Kansas                 241   253    252
Kentucky               262   248    256
Louisiana              226   245    233
Maine                  199   241    218
Maryland               228   233    243
Massachusetts          229   247    246
Michigan               224   233    234
Minnesota              270   262    242
Mississippi            251   236    256
Missouri               251   255    241
Montana                243   249    238
Nebraska               243   256    249
Nevada                 227   240    228
New Hampshire          212   199    226
New Jersey             240   245    249
New Mexico             234   239    241
New York               244   262    242
North Carolina         234   241    226
North Dakota           232   240    254
Ohio                   248   232    259
Oklahoma               219   225    239
Oregon                 258   256    220
Pennsylvania           238   250    223
Rhode Island           238   239    250
South Carolina         231   234    225
South Dakota           254   206    222
Tennessee              262   221    255
Texas                  233   231    252
Utah                   224   227    225
Vermont                223   258    232
Virginia               225   240    228
Washington             230   215    253
West Virginia          228   238    232
Wisconsin              244   223    271
Wyoming                239   246    237

In [ ]:
pivot_df = df.pivot_table(index='HAS INSURANCE', columns='PREFERRED DOCTOR', values='VISITED LAST MONTH', aggfunc='count')
pivot_df


PREFERRED DOCTOR  Dr. Brown  Dr. Johnson  Dr. Jones  Dr. Smith  Dr. Williams
HAS INSURANCE                                                               
No                     3404         3582       3627       3605          3625
Yes                    3563         3544       3593       3636          3649

### Loading Dataset using Polars

In [ ]:
import polars as pl

In [ ]:
df = pl.read_csv('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
df

shape: (105_000, 20)
┌───────────┬────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ Patient   ┆ Gender ┆ City of    ┆ State of   ┆ … ┆ Last Visit ┆ Visit      ┆ Number of  ┆ Prescripti │
│ Age       ┆ ---    ┆ Residence  ┆ Residence  ┆   ┆ (days ago) ┆ Duration   ┆ Tests      ┆ on Cost    │
│ ---       ┆ str    ┆ ---        ┆ ---        ┆   ┆ ---        ┆ (mins)     ┆ ---        ┆ ($)        │
│ str       ┆        ┆ str        ┆ str        ┆   ┆ str        ┆ ---        ┆ str        ┆ ---        │
│           ┆        ┆            ┆            ┆   ┆            ┆ str        ┆            ┆ str        │
╞═══════════╪════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 45        ┆ Other  ┆ West Brian ┆ Kentucky   ┆ … ┆ 193        ┆ 92         ┆ 5          ┆ 15.7175173 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 5532889    │
│ 25        ┆ Female ┆ East Jocel ┆ Colorado   ┆ … ┆ 195        ┆ missing    ┆ 4          ┆ 80.7864728 │
│           ┆        ┆ ynfurt     ┆            ┆   ┆            ┆            ┆            ┆ 4463952    │
│ 51        ┆ Other  ┆ South Lind ┆ Ohio       ┆ … ┆ missing    ┆ 37         ┆ 2          ┆ 64.3921393 │
│           ┆        ┆ seyland    ┆            ┆   ┆            ┆            ┆            ┆ 009105     │
│ 18        ┆ Female ┆ Taylorfort ┆ Pennsylvan ┆ … ┆ 175        ┆ 108        ┆ 2          ┆ 8.87285921 │
│           ┆        ┆            ┆ ia         ┆   ┆            ┆            ┆            ┆ 5315316    │
│ …         ┆ …      ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …          │
│ 37        ┆ Male   ┆ Mckeetown  ┆ Utah       ┆ … ┆ 191        ┆ 47         ┆ 2          ┆ 113.867640 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 25370326   │
│ 51        ┆ Female ┆ Lake Lisac ┆ Nebraska   ┆ … ┆ 172        ┆ 39         ┆ 3          ┆ 35.6224202 │
│           ┆        ┆ hester     ┆            ┆   ┆            ┆            ┆            ┆ 32475356   │
│ missing   ┆ Female ┆ Heidiborou ┆ missing    ┆ … ┆ 180        ┆ 79         ┆ missing    ┆ 124.408340 │
│           ┆        ┆ gh         ┆            ┆   ┆            ┆            ┆            ┆ 9893742    │
│ 45        ┆ Female ┆ South      ┆ Massachuse ┆ … ┆ 167        ┆ 81         ┆ missing    ┆ 87.9106014 │
│           ┆        ┆ Jeffrey    ┆ tts        ┆   ┆            ┆            ┆            ┆ 9353706    │
└───────────┴────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────────┘

In [ ]:
import time

### Comparing Load Times

In [ ]:
start_time = time.time()
pd_df = pd.read_csv('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
pd_time = time.time() - start_time
f"Pandas Execution Time: {pandas_time:.4f} seconds"

'Pandas Execution Time: 0.6733 seconds'

In [ ]:
start_time = time.time()
pl_df = pl.read_csv('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
pl_time = time.time() - start_time
f"Polars Execution Time: {pl_time:.4f} seconds"

'Polars Execution Time: 0.2322 seconds'

##### Polars (0.2322 seconds) has a significantly **faster** load time compared to Pandas (0.6733 seconds) because Polars is faster for big data.